p2panda-rs/src/permissions/key_group/key_group.rs

```diff
@@ -0,0 +1,297 @@
+                            debug!("Skipping {}", author.as_str());
+                        }
+                    }
+                }
+            };
+        }
+
+        // Deduplicate so we have one membership per public key. A membership with more rights
+        // takes precedence here. At the moment that is just memberships that are accepted, vs not
+        // accepted.
+        let mut member_map: HashMap<Author, Vec<Membership>> = HashMap::new();
+        for (author, membership) in member_pool {
+            if let Some(previous_memberships) = member_map.get_mut(author) {
+                previous_memberships.push(membership.clone());
+            } else {
+                member_map.insert(author.clone(), vec![membership.clone()]);
+            }
+        }
+
+        let key_group = KeyGroup {
+            id: key_group.id().clone(),
+            view_id: key_group.view_id().clone(),
+            name: key_group.name().to_string(),
+            members: member_map,
+        };
+
+        Ok(key_group)
+    }
+
+    /// Create a new key group from predecessor documents.
+    ///
+    /// The `documents` parameter should contain documents for the key group itself and all
+    /// (request, response) pairs.
+    ///
+    /// The `member_key_groups` parameter should contain all [`KeyGroup`]s that are members.
+    pub fn new_from_documents(
+        key_group_id: DocumentId,
+        documents: &[Document],
+        member_key_groups: &[KeyGroup],
+    ) -> Result<KeyGroup, KeyGroupError> {
+        let mut key_group = None;
+        let mut requests = HashMap::new();
+        let mut responses = Vec::new();
+
+        for document in documents {
+            match document.schema() {
+                SchemaId::KeyGroupResponse => {
+                    responses.push(MembershipResponseView::try_from(document.clone())?);
+                }
+                SchemaId::KeyGroupRequest => {
+                    let request = MembershipRequestView::try_from(document.clone())?;
+                    if request.key_group() == &key_group_id {
+                        requests.insert(request.view_id().clone(), request);
+                    }
+                }
+                _ => (),
+            }
+            if document.id() == &key_group_id {
+                key_group = Some(KeyGroupView::try_from(document.clone())?);
+            }
+        }
+        if key_group.is_none() {
+            return Err(KeyGroupError::MissingKeyGroupView);
+        }
+
+        let mut members: Vec<Membership> = Vec::new();
+        for response in responses {
+            // Remove requests for which we have a response
+            if let Some(request) = requests.remove(response.request()) {
+                members.push(Membership::from_confirmation(
+                    request.clone(),
+                    Some(response.clone()),
+                )?);
+            };
+        }
+
+        for request in requests.values() {
```

Comment:
Maybe a comment saying these are requests which don't have a response yet would be nice.